arch/x86/pagetables.c

```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+        spin_unlock(&vmap_lock);
+        return 0;
+    }
+
+    tab = init_map_mfn(l2e->mfn);
+    pte_t *l1e = l1_table_entry((pte_t *) tab, va);
+    if (mfn_invalid(l1e->mfn) || !l1e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    set_pgentry(&l1e->entry, MFN_INVALID, PT_NO_FLAGS);
+    spin_unlock(&vmap_lock);
+    return 0;
+}
+
+int vunmap_kern(void *va) {
+    return _vunmap(&cr3, va);
+}
+
+int vunmap_user(void *va) {
+    return _vunmap(&user_cr3, va);
+}
+
+#define MMAP_RANGE_SIZE(size, mfn, flags, mmap_flags)                                    \
+    do {                                                                                 \
+        if (likely((mmap_flags) &MMAP_KERNEL))                                           \
+            kmap_##size((mfn), (flags));                                                 \
+        if ((mmap_flags) &MMAP_IDENT)                                                    \
+            vmap_##size(mfn_to_virt((mfn)), (mfn), (flags));                             \
+        if ((mmap_flags) &MMAP_USER)                                                     \
+            vmap_user_##size(mfn_to_virt_user((mfn)), (mfn), (flags | _PAGE_USER));      \
+        if ((mmap_flags) &MMAP_USER_KERNEL)                                              \
+            vmap_##size(mfn_to_virt_user((mfn)), (mfn), (flags & ~_PAGE_USER));          \
+        if ((mmap_flags) &MMAP_KERNEL_MAP)                                               \
+            vmap_##size(mfn_to_virt_map((mfn)), (mfn), (flags));                         \
+    } while (0)
+
+#define MMAP_RANGE_4K(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(4k, mfn, flags, mmap_flags)
+#define MMAP_RANGE_2M(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(2m, mfn, flags, mmap_flags)
+#define MMAP_RANGE_1G(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(1g, mfn, flags, mmap_flags)
+
+void *mmap_range(paddr_t paddr, size_t size, unsigned long flags,
+                 mmap_flags_t mmap_flags) {
+    paddr_t cur = paddr;
+    paddr_t end = cur + size;
+
+    /* Round up to the next page boundary unless it is page aligned */
+    end = (end + PAGE_SIZE - 1) & PAGE_MASK;
+
+    spin_lock(&mmap_lock);
+
+    for (; (cur % PAGE_SIZE_2M) && cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
+    if (likely(cur >= end))
+        goto done_4k;
+
+    for (; (cur % PAGE_SIZE_1G) && cur + PAGE_SIZE_2M <= end; cur += PAGE_SIZE_2M)
+        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
+
+    if (likely(cur >= end))
+        goto done_2m;
+
+    for (; cur + PAGE_SIZE_1G <= end; cur += PAGE_SIZE_1G)
+        MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
+
+    for (; cur + PAGE_SIZE_2M <= end; cur += PAGE_SIZE_2M)
+        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
+
+done_2m:
+    for (; cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
```

Comment:
Can't these multiple first ascending, than descending for-loops be simplified to a single loop like this?:

```c
while (size >= PAGE_SIZE) {
    if (size >= PAGE_SIZE_1G && !(cur % PAGE_SIZE_1G)) {
        MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
        cur += PAGE_SIZE_1G;
        size -= PAGE_SIZE_1G;
        continue;
    }

    if (size >= PAGE_SIZE_2M && !(cur % PAGE_SIZE_2M)) {
        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
        cur += PAGE_SIZE_2M;
        size -= PAGE_SIZE_2M;
        continue;
    }

    MMAP_RANGE_4k(paddr_to_mfn(cur), flags, mmap_flags);
    cur += PAGE_SIZE;
    size -= PAGE_SIZE;
}
```

This should achieve the same thing (using the best fitting mapping for a given size/alignment) with less code.

Reply:
Ok, good idea... the following seems to work fine:

```c
    while (cur < end) {
        if (size >= PAGE_SIZE_1G && !(cur % PAGE_SIZE_1G)) {
            MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
            cur += PAGE_SIZE_1G;
            size -= PAGE_SIZE_1G;
            continue;
        }

        if (size >= PAGE_SIZE_2M && !(cur % PAGE_SIZE_2M)) {
            MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
            cur += PAGE_SIZE_2M;
            size -= PAGE_SIZE_2M;
            continue;
        }

        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
        cur += PAGE_SIZE;
        size -= PAGE_SIZE;
    }
```

Reply:
Should be equivalent to what I proposed, yes.